### PR TITLE
NODE-2528 Typed errors replacing exceptions in RIDE sources

### PR DESCRIPTION
**Pull request by xrtm000**

### CODE REVIEW

**Review by ivan-mashonskiy — APPROVED** [review]

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala [hunk 1]
```diff
@@ -434,16 +441,24 @@ object Functions {
...
+                  r => env.accountBalanceOf(r, Some(assetId.arr)).map(_.map(CONST_LONG).leftMap(CommonError(_)))
+                )
             case xs =>
               notImplemented[F, EVALUATED](s"assetBalance(a: Address|Alias, u: ByteVector|Unit)", xs)
           }
       }
     }
 
-  val assetBalanceV4F: BaseFunction[Environment] =
+  def assetBalanceV4F(typedError: Boolean): BaseFunction[Environment] =
```
Comment (vsuharnikov) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala]: We couldn't add `typedError` argument everywhere, if we've had the `ThrownError` subclass of `ExecutionError`. E.g.:
1. Return `ThrownError` if we expect an exception. The class may contain this exception, so we don't lose a stacktrace.
2. Check in the end the `typedError` flag and throw the exception if it is needed.

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala [hunk 2]
```diff
@@ -434,16 +441,24 @@ object Functions {
         override def evaluate[F[_]: Monad](env: Environment[F], args: List[EVALUATED]): F[Either[ExecutionError, EVALUATED]] =
           args match {
             case (c: CaseObj) :: u :: Nil if u == unit =>
-              env.accountBalanceOf(caseObjToRecipient(c), None).map(_.map(CONST_LONG).leftMap(CommonError(_)))
+              caseObjToRecipient(c, typedError)
+                .fold(
+                  _.asLeft[EVALUATED].pure[F],
```
Comment (vsuharnikov) [inline lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala]: Should we make `caseObjToRecipient` return `F[Either[ExecutionError, Recipient]]` and handle this error? We convert the result to `F` anyway now.

Comment (xrtm000) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala]: `.fold` will become unavailable for `F[Either[*, *]]`

File: lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala (hunk 1 above)
Comment (xrtm000) [reply lang/shared/src/main/scala/com/wavesplatform/lang/v1/evaluator/ctx/impl/waves/Functions.scala]: Done

**Review by vsuharnikov — APPROVED** [review]

**Review by DrBlast — APPROVED** [review]

**Review by phearnot — APPROVED** [review]

**Pull request merged**